### ISSUE TITLE
Nvidia CRI doesn't apply with imports in config.toml [containerd]

### ISSUE THREAD

**Issue by nneram** [NVIDIA/gpu-operator#301]
### 1. Nvidia CRI doesn't apply with `imports = ["/etc/containerd/config.d/*.toml"]` in the `/etc/containerd/config.toml`

In my cluster I add some private registries with an import in the config.toml but nvidia CRI does not apply with it.
If I remove `imports = ["/etc/containerd/config.d/*.toml"]` from the file it works well.

### 2. Steps to reproduce the issue

Add `imports = ["/etc/containerd/config.d/*.toml"]` to `/etc/containerd/config.toml` and add file like `/etc/containerd/config.d/registries.toml` to add some configurations. 

### 3. Information to [attach](https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/) (optional if deemed irrelevant)

 - [ ] Containerd config: `kubectl cat /etc/containerd/config.toml`
```
disabled_plugins = []
imports = ["/etc/containerd/config.d/*.toml"]
oom_score = 0
plugin_dir = ""
required_plugins = []
root = "/var/lib/containerd"
state = "/run/containerd"
version = 2

[cgroup]
  path = ""

[debug]
  address = ""
  gid = 0
  level = ""
  uid = 0

[grpc]
  address = "/run/containerd/containerd.sock"
  gid = 0
  max_recv_message_size = 16777216
  max_send_message_size = 16777216
  tcp_address = ""
  tcp_tls_cert = ""
  tcp_tls_key = ""
  uid = 0

[metrics]
  address = ""
  grpc_histogram = false

[plugins]

  [plugins."io.containerd.gc.v1.scheduler"]
    deletion_threshold = 0
    mutation_threshold = 100
    pause_threshold = 0.02
    schedule_delay = "0s"
    startup_delay = "100ms"

  [plugins."io.containerd.grpc.v1.cri"]
    disable_apparmor = false
    disable_cgroup = false
    disable_hugetlb_controller = true
    disable_proc_mount = false
    disable_tcp_service = true
    enable_selinux = false
    enable_tls_streaming = false
    ignore_image_defined_volumes = false
    max_concurrent_downloads = 3
    max_container_log_line_size = 16384
    restrict_oom_score_adj = false
    sandbox_image = "k8s.gcr.io/pause:3.2"
    selinux_category_range = 1024
    stats_collect_period = 10
    stream_idle_timeout = "4h0m0s"
    stream_server_address = "127.0.0.1"
    stream_server_port = "0"
    systemd_cgroup = false
    tolerate_missing_hugetlb_controller = true
    unset_seccomp_profile = ""

    [plugins."io.containerd.grpc.v1.cri".cni]
      bin_dir = "/opt/cni/bin"
      conf_dir = "/etc/cni/net.d"
      conf_template = ""
      max_conf_num = 1

    [plugins."io.containerd.grpc.v1.cri".containerd]
      default_runtime_name = "nvidia"
      disable_snapshot_annotations = true
      discard_unpacked_layers = false
      no_pivot = false
      snapshotter = "overlayfs"

      [plugins."io.containerd.grpc.v1.cri".containerd.default_runtime]
        base_runtime_spec = ""
        privileged_without_host_devices = false
        runtime_engine = ""
        runtime_root = ""
        runtime_type = ""

      [plugins."io.containerd.grpc.v1.cri".containerd.runtimes]

        [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.nvidia]
          base_runtime_spec = ""
          privileged_without_host_devices = false
          runtime_engine = ""
          runtime_root = ""
          runtime_type = "io.containerd.runc.v2"

          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.nvidia.options]
            BinaryName = "/usr/local/nvidia/toolkit/nvidia-container-runtime"
            SystemdCgroup = true

        [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.nvidia-experimental]
          base_runtime_spec = ""
          privileged_without_host_devices = false
          runtime_engine = ""
          runtime_root = ""
          runtime_type = "io.containerd.runc.v2"

          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.nvidia-experimental.options]
            BinaryName = "/usr/local/nvidia/toolkit/nvidia-container-runtime-experimental"
            SystemdCgroup = true

        [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
          base_runtime_spec = ""
          privileged_without_host_devices = false
          runtime_engine = ""
          runtime_root = ""
          runtime_type = "io.containerd.runc.v2"

          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
            SystemdCgroup = true

      [plugins."io.containerd.grpc.v1.cri".containerd.untrusted_workload_runtime]
        base_runtime_spec = ""
        privileged_without_host_devices = false
        runtime_engine = ""
        runtime_root = ""
        runtime_type = ""

    [plugins."io.containerd.grpc.v1.cri".image_decryption]
      key_model = ""

    [plugins."io.containerd.grpc.v1.cri".registry]

      [plugins."io.containerd.grpc.v1.cri".registry.mirrors]

        [plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.io"]
          endpoint = ["https://registry-1.docker.io"]

    [plugins."io.containerd.grpc.v1.cri".x509_key_pair_streaming]
      tls_cert_file = ""
      tls_key_file = ""

  [plugins."io.containerd.internal.v1.opt"]
    path = "/opt/containerd"

  [plugins."io.containerd.internal.v1.restart"]
    interval = "10s"

  [plugins."io.containerd.metadata.v1.bolt"]
    content_sharing_policy = "shared"

  [plugins."io.containerd.monitor.v1.cgroups"]
    no_prometheus = false

  [plugins."io.containerd.runtime.v1.linux"]
    no_shim = false
    runtime = "runc"
    runtime_root = ""
    shim = "containerd-shim"
    shim_debug = false

  [plugins."io.containerd.runtime.v2.task"]
    platforms = ["linux/amd64"]

  [plugins."io.containerd.service.v1.diff-service"]
    default = ["walking"]

  [plugins."io.containerd.snapshotter.v1.devmapper"]
    async_remove = false
    base_image_size = ""
    pool_name = ""
    root_path = ""

[timeouts]
  "io.containerd.timeout.shim.cleanup" = "5s"
  "io.containerd.timeout.shim.load" = "5s"
  "io.containerd.timeout.shim.shutdown" = "3s"
  "io.containerd.timeout.task.state" = "2s"

[ttrpc]
  address = ""
  gid = 0
  uid = 0
```
 - [ ] kubectl get -n gpu-operator-resources pods | grep Init
```
gpu-feature-discovery-cgg9s                0/1     Init:0/1    0          23m
nvidia-dcgm-bh7mz                          0/1     Init:0/1    0          23m
nvidia-dcgm-exporter-scjbm                 0/1     Init:0/1    0          23m
nvidia-device-plugin-daemonset-fd7qv       0/1     Init:0/1    0          23m
nvidia-operator-validator-zv5h9            0/1     Init:0/4    0          23m
```
 - [ ] If a pod/ds is in an error state or pending state `kubectl describe pod -n gpu-operator-resources`
```
  Warning  FailedCreatePodSandBox  5s (x7 over 84s)    kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to get sandbox runtime: no runtime for "nvidia" is configured
```
 - [ ] All it's configured well acording to this : 

```
toolkit:
   env:
   - name: CONTAINERD_CONFIG
   value: /etc/containerd/config.toml
   - name: CONTAINERD_SOCKET
   value: /run/containerd/containerd.sock
   - name: CONTAINERD_RUNTIME_CLASS
   value: nvidia
   - name: CONTAINERD_SET_AS_DEFAULT
   value: true
```

 - [ ] Versions: 
 ```
 Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.2", GitCommit:"8b5a19147530eaac9476b0ab82980b4088bbc1b2", GitTreeState:"clean", BuildDate:"2021-09-15T21:38:50Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.2", GitCommit:"f5743093fd1c663cb0cbc89748f730662345d44d", GitTreeState:"clean", BuildDate:"2020-09-16T13:32:58Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```

Nvidia GPU Operator: 
```
 GPU Operator 1.8.2 Release
```

**Comment by shivamerla** [comment]
@nneram Thanks for reporting this. Will check and fix this.

**Comment by elezar** [comment]
@nneram which version of `containerd` is being used on this system? What other files are present at `/etc/containerd/config.d/*.toml`? Do any of these contain any `nvidia`-specific configs?

**Comment by nneram** [comment]
Hi @elezar, 

Thank you for pointing out this issue, in deed their is just one file in `/etc/containerd/config.d/*.toml`: 
```
cat registries.toml

    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker-virtual.repository.com"]
    endpoint = ["http://docker-virtual.repository.com"]
    [plugins."io.containerd.grpc.v1.cri".registry.configs."docker-virtual.repository.com".tls]
    insecure_skip_verify = true
[plugins."io.containerd.grpc.v1.cri".registry.configs."docker-virtual.repository.com".auth]
     username="svc-prj"
     password="xxxx
```

My version of `containerd` is `1.4.12` and I guess the default CRI is taking into account.  

I will try to set this:

**Comment by elezar** [comment]
Yes, this does seem to be a limitation to how imports are implemented in containerd. Note that the code I was looking at was at `HEAD` and seemed to have the same issue, so I don't think this is limited to `containerd` `1.4.12`.

Considering your registries file, my understanding of what is happening is that the `plugins."io.containerd.grpc.v1.cri"` entry (containing only the `registriy` config) is replacing the entire `plugins."io.containerd.grpc.v1.cri"` config from the root file (see the code [here](https://github.com/containerd/containerd/blob/63b7e5771e8914f3c36c707f3b5fc4846b11997b/services/server/config/config.go#L331-L334)).

**Comment by nneram** [comment]
I tryed this:
```
toolkit:
   env:
   - name: CONTAINERD_CONFIG
   value: /etc/containerd/config.d/repositories.toml
```

But I ran into a problem, both config are not modified either in `/etc/containerd/config.d/repositories.toml` or `/etc/containerd/config.toml` after the helm install.

I was looking for logs but if I didn't found any relevant, which pod is adding the conf in the `.toml` ? 
I looked in the doc and my guessing are that it's this one: 
```
nvidia-container-toolkit-ctr  ●  nvcr.io/nvidia/k8s/container-toolkit:1.7.1-ubuntu18.04     false  PodInitializing
``` 

But as you can see it's not initializing. I looked in issue and I found this: 
https://github.com/NVIDIA/gpu-operator/issues/297

So i guess that the operator installed the toolkit the first time and when I deleted it to reinstall it's not working anymore. 
I deleted the `/usr/local/nvidia` on my node but I still have the problem. 

Do you have a clue to resolve this ? I guess I can walkarround the first problem but this one I'm stuck.

**Comment by shivamerla** [comment]
@nneram I think the suggestion @elezar is that your imported file is overwriting the entire `plugins."io.containerd.grpc.v1.cri` config in main `/etc/containerd/config.toml` file due to what appears to be an issues in containerd. So, you would need to merge these two files into /etc/containerd/config.toml instead of specifying the import file as containerd config.

**Comment by nneram** [comment]
@shivamerla Yeah it's the idea so I changed the toolkit configuration with this target `/etc/containerd/config.d/repositories.toml`. Now CRI plugin conf are on the same file that will be import (so NVDIA will be) and it worked. The  `crictl info` showed NVIDIA as expected. I did that because I don't have the possibility to merge them because I don't have the hand on it. Never the less it worked as expected but I got this new problem :
I wanted to reinstall the all thing but now the `/etc/containerd/config.toml` is not update at all even with the default value for the toolkit. I have no clue why the `config.toml` is not updated and `containerd` not restarted but is suspect the `toolkit` pod.
